mtp_api/apps/transaction/tests/test_views.py

```diff
@@ -7,105 +7,744 @@
-            # we have to do if Prison because some transactions don't have a prison attached
-            # and therefore no user will be able to see those...
-            # TODO: figure out who is responsible for these transactions
-            if prison:
-                user = get_user_for_prison(prison)
-                self.client.force_authenticate(
-                    user=user)
-                response = self.client.get(url, format='json')
-                self.assertEqual(response.status_code, status.HTTP_200_OK)
-                self.assertEqual(response.data['count'], count)
-                self.assertResultsBelongToUser(response.data['results'], user)
-                if count > api_settings.PAGE_SIZE:
-                    self.assertNotEqual(response.data['next'], None)
-
-                # get transaction from response data and make sure that they
-                # all belong to last upload
-                self.assertTransactionsInUpload(response, 2)
-
-    def test_list_specific_upload(self):
-        upload_counter = 1
-        url = '{url}?upload_counter={upload_counter}'.\
-            format(
-                url=reverse('transaction-list'),
-                upload_counter=upload_counter
+            self.assertEqual(response.status_code, status.HTTP_200_OK)
+            self.assertEqual(response.data['count'], len(expected_ids))
+
+            self.assertListEqual(
+                sorted([t['id'] for t in response.data['results']]),
+                sorted(expected_ids)
             )
 
-        unique_prisons = self.calculate_prison_transaction_counts(
-            upload_counter=upload_counter)
-        for prison, count in unique_prisons.items():
-            if prison:
-                user = get_user_for_prison(prison)
-                self.client.force_authenticate(
-                    user=user)
+    def _get_list_url(self, prison, status=None):
+        url = reverse(
+            'transaction-prison-list', kwargs={
+                'prison_id': prison.pk
+            }
+        )
+
+        params = {
+            'limit': 10000
+        }
+        if status:
+            params['status'] = status
+
+        return '{url}?{params}'.format(
+            url=url, params=urllib.parse.urlencode(params)
+        )
+
+    def test_fails_with_logged_in_managing_different_prison(self):
+        logged_in_user_prison = self.prisons[0]
+        other_prison = self.prisons[1]
+
+        logged_in_user = get_users_for_prison(logged_in_user_prison)[0]
+
+        url = self._get_list_url(other_prison)
+
+        self.client.force_authenticate(user=logged_in_user)
+
+        response = self.client.get(url, format='json')
+
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_all(self):
+        """
+        GET without params should return all transactions linked
+        to that prison.
+        """
+        self._request_and_assert()
```

Comment:
@SteveMarshall Yes, I started like that but they were getting very long, repetitive and complicated so I thought this looked better.
Any suggestions on how to improve it? Otherwise, I'd rather keep it like this.